File: backend/pdf_gen.py
"""Accounting PDF generation with reportlab."""
import io
from datetime import datetime
from reportlab.lib.pagesizes import A4
from reportlab.lib import colors
from reportlab.lib.styles import getSampleStyleSheet, ParagraphStyle
from reportlab.lib.units import cm
from reportlab.platypus import SimpleDocTemplate, Paragraph, Spacer, Table, TableStyle


BRAND = colors.HexColor("#7FA9A8")
DARK = colors.HexColor("#1A1C18")


def build_accounting_pdf(month: int, year: int, orders: list, restaurant_name: str) -> bytes:
    buf = io.BytesIO()
    doc = SimpleDocTemplate(buf, pagesize=A4, leftMargin=2 * cm, rightMargin=2 * cm, topMargin=2 * cm, bottomMargin=2 * cm)
    styles = getSampleStyleSheet()
    title_style = ParagraphStyle("t", parent=styles["Heading1"], textColor=DARK, fontSize=22, spaceAfter=6)
    sub_style = ParagraphStyle("s", parent=styles["Normal"], textColor=colors.HexColor("#5C6057"), fontSize=10, spaceAfter=18)
    h2 = ParagraphStyle("h2", parent=styles["Heading2"], textColor=BRAND, fontSize=13, spaceBefore=16, spaceAfter=8)

    story = []
    story.append(Paragraph(f"{restaurant_name} — Comptabilité", title_style))
    story.append(Paragraph(f"Période : {month:02d}/{year} · Généré le {datetime.now().strftime('%d.%m.%Y %H:%M')}", sub_style))

    # Totals by VAT rate
    net_by_vat: dict = {}
    vat_by_vat: dict = {}
    gross = 0.0
    product_qty: dict = {}

    for o in orders:
        rate = o.get("vat_rate", 0.026)
        net_by_vat[rate] = net_by_vat.get(rate, 0.0) + o.get("subtotal", 0.0) - o.get("discount_amount", 0.0)
        vat_by_vat[rate] = vat_by_vat.get(rate, 0.0) + o.get("vat_amount", 0.0)
        gross += o.get("total", 0.0)
        for it in o.get("items", []):
            product_qty[it["name"]] = product_qty.get(it["name"], 0) + it["quantity"]

    net = sum(net_by_vat.values())
    vat_total = sum(vat_by_vat.values())

    story.append(Paragraph("Synthèse", h2))
    summary = [
        ["Chiffre d'affaires brut (TTC)", f"CHF {gross:.2f}"],
        ["CA net (HT)", f"CHF {net:.2f}"],
        ["TVA totale", f"CHF {vat_total:.2f}"],
        ["Nombre de commandes", str(len(orders))],
    ]
    t = Table(summary, colWidths=[9 * cm, 7 * cm])
    t.setStyle(TableStyle([
        ("BACKGROUND", (0, 0), (-1, 0), colors.HexColor("#F5F2EA")),
        ("TEXTCOLOR", (0, 0), (-1, -1), DARK),
        ("FONTSIZE", (0, 0), (-1, -1), 10),
        ("BOTTOMPADDING", (0, 0), (-1, -1), 8),
        ("TOPPADDING", (0, 0), (-1, -1), 8),
        ("LINEBELOW", (0, 0), (-1, -1), 0.3, colors.HexColor("#DDDDDD")),
    ]))
    story.append(t)

    story.append(Paragraph("TVA détaillée par taux", h2))
    vat_rows = [["Taux", "Base HT", "TVA collectée"]]
    for rate in sorted(net_by_vat.keys()):
        vat_rows.append([f"{rate*100:.1f}%", f"CHF {net_by_vat[rate]:.2f}", f"CHF {vat_by_vat[rate]:.2f}"])
    tv = Table(vat_rows, colWidths=[5 * cm, 5.5 * cm, 5.5 * cm])
    tv.setStyle(TableStyle([
        ("BACKGROUND", (0, 0), (-1, 0), BRAND),
        ("TEXTCOLOR", (0, 0), (-1, 0), colors.white),
        ("FONTSIZE", (0, 0), (-1, -1), 10),
        ("BOTTOMPADDING", (0, 0), (-1, -1), 8),
        ("TOPPADDING", (0, 0), (-1, -1), 8),
        ("GRID", (0, 0), (-1, -1), 0.3, colors.HexColor("#DDDDDD")),
    ]))
    story.append(tv)

    story.append(Paragraph("Produits vendus (par quantité)", h2))
    prods = sorted(product_qty.items(), key=lambda x: -x[1])
    prod_rows = [["Produit", "Quantité"]]
    for name, qty in prods:
        prod_rows.append([name, str(qty)])
    if len(prod_rows) == 1:
        prod_rows.append(["—", "0"])
    tp = Table(prod_rows, colWidths=[12 * cm, 4 * cm])
    tp.setStyle(TableStyle([
        ("BACKGROUND", (0, 0), (-1, 0), BRAND),
        ("TEXTCOLOR", (0, 0), (-1, 0), colors.white),
        ("FONTSIZE", (0, 0), (-1, -1), 10),
        ("BOTTOMPADDING", (0, 0), (-1, -1), 6),
        ("TOPPADDING", (0, 0), (-1, -1), 6),
        ("GRID", (0, 0), (-1, -1), 0.3, colors.HexColor("#DDDDDD")),
    ]))
    story.append(tp)

    story.append(Spacer(1, 1 * cm))
    story.append(Paragraph("Document anonymisé — commandes supprimées exclues.", sub_style))

    doc.build(story)
    return buf.getvalue()

File: backend/pdfGen.js
/** Accounting PDF via PDFKit — port of backend/pdf_gen.py */
const PDFDocument = require("pdfkit");

const BRAND = "#7FA9A8";
const DARK = "#1A1C18";
const MUTED = "#5C6057";
const CREAM = "#F5F2EA";
const GRID = "#DDDDDD";

function money(v) { return `CHF ${Number(v || 0).toFixed(2)}`; }

/**
 * @param {number} month
 * @param {number} year
 * @param {Array<object>} orders
 * @param {string} restaurantName
 * @returns {Promise<Buffer>}
 */
function buildAccountingPdf(month, year, orders, restaurantName) {
  return new Promise((resolve, reject) => {
    const doc = new PDFDocument({ size: "A4", margins: { top: 56, bottom: 56, left: 56, right: 56 } });
    const chunks = [];
    doc.on("data", c => chunks.push(c));
    doc.on("end", () => resolve(Buffer.concat(chunks)));
    doc.on("error", reject);

    // ---- Header ----
    doc.fillColor(DARK).fontSize(22).text(`${restaurantName} — Comptabilité`, { align: "left" });
    doc.moveDown(0.2);
    doc.fillColor(MUTED).fontSize(10).text(
      `Période : ${String(month).padStart(2, "0")}/${year} · Généré le ${new Date().toLocaleString("fr-CH", { timeZone: "Europe/Zurich" })}`,
      { align: "left" }
    );
    doc.moveDown(1.2);

    // ---- Compute totals ----
    const netByVat = {}, vatByVat = {};
    let gross = 0;
    const productQty = {};
    for (const o of orders) {
      const rate = Number(o.vat_rate ?? 0.026);
      netByVat[rate] = (netByVat[rate] || 0) + (Number(o.subtotal) || 0) - (Number(o.discount_amount) || 0);
      vatByVat[rate] = (vatByVat[rate] || 0) + (Number(o.vat_amount) || 0);
      gross += Number(o.total) || 0;
      for (const it of o.items || []) {
        productQty[it.name] = (productQty[it.name] || 0) + (Number(it.quantity) || 0);
      }
    }
    const net = Object.values(netByVat).reduce((a, b) => a + b, 0);
    const vatTotal = Object.values(vatByVat).reduce((a, b) => a + b, 0);

    // ---- Section: Synthèse ----
    doc.fillColor(BRAND).fontSize(13).text("Synthèse");
    doc.moveDown(0.4);
    const summary = [
      ["Chiffre d'affaires brut (TTC)", money(gross)],
      ["CA net (HT)", money(net)],
      ["TVA totale", money(vatTotal)],
      ["Nombre de commandes", String(orders.length)],
    ];
    renderTable(doc, summary, [260, 200], { header: false });

    doc.moveDown(0.8);
    doc.fillColor(BRAND).fontSize(13).text("TVA détaillée par taux");
    doc.moveDown(0.4);
    const vatRows = [["Taux", "Base HT", "TVA collectée"]];
    for (const rate of Object.keys(netByVat).map(Number).sort((a, b) => a - b)) {
      vatRows.push([`${(rate * 100).toFixed(1)}%`, money(netByVat[rate]), money(vatByVat[rate])]);
    }
    renderTable(doc, vatRows, [153, 155, 155], { header: true });

    doc.moveDown(0.8);
    doc.fillColor(BRAND).fontSize(13).text("Produits vendus (par quantité)");
    doc.moveDown(0.4);
    const prodRows = [["Produit", "Quantité"]];
    const sorted = Object.entries(productQty).sort((a, b) => b[1] - a[1]);
    for (const [name, qty] of sorted) prodRows.push([name, String(qty)]);
    if (prodRows.length === 1) prodRows.push(["—", "0"]);
    renderTable(doc, prodRows, [340, 120], { header: true });

    doc.moveDown(1.2);
    doc.fillColor(MUTED).fontSize(9).text("Document anonymisé — commandes supprimées exclues.", { align: "left" });

    doc.end();
  });
}

/** Simple table renderer. rows is a 2D array. First row is header if opts.header. */
function renderTable(doc, rows, colWidths, opts = {}) {
  const startX = doc.page.margins.left;
  let y = doc.y;
  const rowH = 22;
  const totalW = colWidths.reduce((a, b) => a + b, 0);

  rows.forEach((row, i) => {
    if (y + rowH > doc.page.height - doc.page.margins.bottom) {
      doc.addPage();
      y = doc.page.margins.top;
    }
    const isHeader = opts.header && i === 0;
    if (isHeader) doc.rect(startX, y, totalW, rowH).fill(BRAND);
    else if (!opts.header && i === 0) doc.rect(startX, y, totalW, rowH).fill(CREAM);
    else doc.rect(startX, y, totalW, rowH).fill("#FFFFFF");

    let x = startX;
    row.forEach((cell, ci) => {
      const w = colWidths[ci];
      doc.fillColor(isHeader ? "#FFFFFF" : DARK).fontSize(10).text(String(cell), x + 6, y + 6, { width: w - 12, height: rowH - 8, ellipsis: true });
      x += w;
    });

    // grid
    doc.lineWidth(0.3).strokeColor(GRID).rect(startX, y, totalW, rowH).stroke();
    y += rowH;
  });
  doc.y = y;
}

module.exports = { buildAccountingPdf };
